## Выполнили: Жерноклеев Дмитрий Дамирович, Индык Илья Андреевич, Реентович Александр Анатольевич

# Домашнее задание
## Harry Potter and the Action Prediction Challenge from Natural Language

*deadline*: 14 ноября 2022, 23:59

В этом домашнем задании вы будете работать с корпусом Harry Potter and the Action Prediction Challenge. Корпус собран из фанфиков о Гарри Поттере и состоит из двух частей: 1) сырые тексты, 2) фрагменты текстов, описывающих ситуацию, в которой произнесено заклинание.

Корпус описан в статье: https://arxiv.org/pdf/1905.11037.pdf

David Vilares and Carlos Gómez-Rodríguez. Harry Potter and the Action Prediction Challenge from Natural Language. 2019 Annual Conference of the North American Chapter of the Association for Computational Linguistics. To appear.

Код для сбора корпуса находится в репозитории: https://github.com/aghie/hpac . Корпус можно скачать по инструкции из этого репозитория, но для экономии времени авторы задания уже скачали и подготовили данные к работе. 

Ссылки на собранный корпус: 
* Сырые тексты:  https://www.dropbox.com/s/23xet9kvbqna1qs/hpac_raw.zip?dl=0
* Токенизированные тексты в нижнем регистре: https://www.dropbox.com/s/gwfgmomdbetvdye/hpac_lower_tokenized.zip?dl=0
* train-test-dev: https://www.dropbox.com/s/3vdz0mouvex8abd/hpac_splits.zip?dl=0

Части 1, 2 задания должны быть выполнены на полных текстах (сырых или предобработанных -- на ваше усмотрение), Часть 3 – на разбиение на тестовое, отладочное и обучающее множества. Тестовое множество должно быть использовано исключительно для тестирования моделей, обучающее и отладочное – для выбора модели и параметров. 

В статье и репозитории вы найдете идеи, которые помогут вам выполнить домашнее задание. Их стоит воспринимать как руководство к действию, и не стоит их копировать и переиспользовать. Обученные модели использовать не нужно, код для их обучения можно использовать как подсказку. 

## ПРАВИЛА
1. Домашнее задание выполняется в группе до 3-х человек.
2. Домашнее задание оформляется в виде отчета либо в .pdf файле, либо ipython-тетрадке. 
3. Отчет должен содержать: нумерацию заданий и пунктов, которые вы выполнили, код решения, и понятное пошаговое описание того, что вы сделали. Отчет должен быть написан в академическом стиле, без излишнего использования сленга и с соблюдением норм русского языка.
4. Не стоит копировать фрагменты лекций, статей и Википедии в ваш отчет.
5. Отчеты, состоящие исключительно из кода, не будут проверены и будут автоматически оценены нулевой оценкой.
6. Плагиат и любое недобросоветсное цитирование приводит к обнуление оценки. 


## Часть 1. [2 балла] Эксплоративный анализ 
1. Найдите топ-1000 слов по частоте без учета стоп-слов.
2. Найдите топ-10 по частоте: имен, пар имя + фамилия, пар вида ''профессор'' + имя / фамилия. 

[бонус] Постройте тематическую модель по корпусу HPAC.

[бонус] Найдите еще что-то интересное в корпусе (что-то специфичное для фанфиков или фентези-тематики)

## Часть 2. [2 балла] Модели представления слов 
Обучите модель представления слов (word2vec, GloVe, fastText или любую другую) на материале корпуса HPAC.
1. Продемонстрируйте, как работает поиск синонимов, ассоциаций, лишних слов в обученной модели. 
2. Визуализируйте топ-1000 слов по частоте без учета стоп-слов (п. 1.1) с помощью TSNE или UMAP (https://umap-learn.readthedocs.io).

## Часть 3. [5 баллов] Классификация текстов
Задача классификации формулируется так: данный фрагмент фанфика описывают какую-то ситуацию, которая предшествует произнесению заклинания. Требуется по тексту предсказать, какое именно заклинание будет произнесено. Таким образом, заклинание - это фактически метка класса. Основная мера качества – macro $F_1$.
Обучите несколько классификаторов и сравните их между собой. Оцените качество классификаторов на частых и редких классах. Какие классы чаще всего оказываются перепутаны? Связаны ли ошибки со смыслом заклинаний?

Используйте фрагменты из множества train для обучения, из множества dev для отладки, из множества test – для тестирования и получения итоговых результатов. 

1. [1 балл] Используйте fastText в качестве baseline-классификатора.
2. [2 балла] Используйте сверточные  или реккурентные сети в качестве более продвинутого классификатора. Поэкспериментируйте с количеством и размерностью фильтров, используйте разные размеры окон, попробуйте использовать $k$-max pooling. 
3. [2 балла] Попробуйте расширить обучающее множество за счет аугментации данных. Если вам понадобится словарь синонимов, можно использовать WordNet (ниже вы найдете примеры).

[бонус] Используйте результат max pooling'а как эмбеддинг входного текста. Визуализируйте эмбеддинги 500-1000 предложений из обучающего множества и изучите свойства получившегося пространства.

[бонус] Используйте ваш любимый классификатор и любые (честные) способы повышения качества классификации и получите macro $F_1$ больше 0.5.

## Часть 4. [1 балл] Итоги
Напишите краткое резюме проделанной работы. Читали ли вы сами Гарри Поттера или фанфики о нем и помогло ли вам знание предметной области в выполнении домашнего задания?

### Данные
Сырые тексты 

In [10]:
!unzip hpac_splits.zip

Archive:  hpac_splits.zip
   creating: hpac_corpus/
  inflating: hpac_corpus/hpac_training_128.tsv  
   creating: __MACOSX/
   creating: __MACOSX/hpac_corpus/
  inflating: __MACOSX/hpac_corpus/._hpac_training_128.tsv  
  inflating: hpac_corpus/hpac_dev_128.tsv  
  inflating: __MACOSX/hpac_corpus/._hpac_dev_128.tsv  
  inflating: hpac_corpus/hpac_test_128.tsv  
  inflating: __MACOSX/hpac_corpus/._hpac_test_128.tsv  
  inflating: __MACOSX/._hpac_corpus  


In [11]:
!unzip w2v_model.zip

Archive:  w2v_model.zip
  inflating: hpac_w2v.model          
  inflating: hpac_w2v.model.syn1neg.npy  
  inflating: hpac_w2v.model.wv.vectors.npy  


train, test, dev файлы

In [1]:
!pip install gensim==4.2.0

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


Выпоним предобрабоку данных и уберем те сэмплы, для которых в нашей word2vec модели нет эмбеддингов (тескт состоит из стоп слов или заклинаний; очевидно, одна фраза из диалога попала в текст для обучения, а следующая - в метку класса данного текста).

In [2]:
import pandas as pd
df = pd.read_csv('hpac_corpus/hpac_training_128.tsv', sep = '\t', header = None)
df_test = pd.read_csv('hpac_corpus/hpac_test_128.tsv', sep = '\t', header = None)
df_dev = pd.read_csv('hpac_corpus/hpac_dev_128.tsv', sep = '\t', header = None)

In [3]:
from gensim.utils import simple_preprocess
df.iloc[:, 2] = df.iloc[:, 2].apply(lambda x: ' '.join(simple_preprocess(x)))
df.iloc[:, 1] = df.iloc[:, 1].apply(lambda x: '__label__' + x)

df_test.iloc[:, 2] = df_test.iloc[:, 2].apply(lambda x: ' '.join(simple_preprocess(x)))
df_test.iloc[:, 1] = df_test.iloc[:, 1].apply(lambda x: '__label__' + x)

df_dev.iloc[:, 2] = df_dev.iloc[:, 2].apply(lambda x: ' '.join(simple_preprocess(x)))
df_dev.iloc[:, 1] = df_dev.iloc[:, 1].apply(lambda x: '__label__' + x)

In [4]:
import numpy as np
from gensim.models import Word2Vec
class Text2VecTransformer:
    def __init__(self, model_dir='./'):
        model = Word2Vec.load(model_dir + "hpac_w2v.model")
        self.model = model
        
    def transform(self, text: str):
        vecs = []
        for word in text.split():
            try:
                word_embedding = self.model.wv[word]
                vecs.append(word_embedding)
            except KeyError:  # если слова нет в словаре w2v, то пропускаем его
                continue
                
        vecs = np.array(vecs)
        return self.aggregation(vecs)
        
    def aggregation(self, vecs: np.array):
        return np.mean(vecs, axis=0)
#         return np.max(vecs, axis=0)  # max-pool ?
    
    def text_summary(self, text: str):
        """Для проверки на здравый смысл и поиска лучшей функции агрегации"""
        return model.wv.similar_by_vector(self.transform(text))

In [5]:
# Remove samples with no embeddings
vectorizer = Text2VecTransformer()
mask_train = df.iloc[:, 2].apply(lambda x: vectorizer.transform(x).sum())
mask_test = df_test.iloc[:, 2].apply(lambda x: vectorizer.transform(x).sum())
mask_dev = df_dev.iloc[:, 2].apply(lambda x: vectorizer.transform(x).sum())

/usr/local/lib/python3.7/dist-packages/numpy/core/fromnumeric.py:3441: RuntimeWarning: Mean of empty slice.
  out=out, **kwargs)


In [6]:
df = df.drop(df[np.isnan(mask_train)].index).reset_index(drop=True)
df_test = df_test.drop(df_test[np.isnan(mask_test)].index).reset_index(drop=True)
df_dev = df_dev.drop(df_dev[np.isnan(mask_dev)].index).reset_index(drop=True)

# Часть 3
## 1. Fasttext

In [178]:
!wget https://github.com/facebookresearch/fastText/archive/v0.9.2.zip
!unzip v0.9.2.zip
%cd fastText-0.9.2
!make

--2022-11-17 14:35:18--  https://github.com/facebookresearch/fastText/archive/v0.9.2.zip
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://codeload.github.com/facebookresearch/fastText/zip/refs/tags/v0.9.2 [following]
--2022-11-17 14:35:18--  https://codeload.github.com/facebookresearch/fastText/zip/refs/tags/v0.9.2
Resolving codeload.github.com (codeload.github.com)... 140.82.121.9
Connecting to codeload.github.com (codeload.github.com)|140.82.121.9|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘v0.9.2.zip’

v0.9.2.zip              [  <=>               ]   4.17M  11.3MB/s    in 0.4s    

2022-11-17 14:35:19 (11.3 MB/s) - ‘v0.9.2.zip’ saved [4369852]

Archive:  v0.9.2.zip
5b5943c118b0ec5fb9cd8d20587de2b2d3966dfe
   creating: fastText-0.9.2/
   creating: fastText-0.9.2/.circleci/
  i

In [179]:
!pip install fasttext

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 68 kB 6.3 MB/s 
  Using cached pybind11-2.10.1-py3-none-any.whl (216 kB)
  Created wheel for fasttext: filename=fasttext-0.9.2-cp37-cp37m-linux_x86_64.whl size=3159017 sha256=4db4a411d9449c266de9740304f76af3b450479155a7965afbd3720f5ed74356
  Stored in directory: /root/.cache/pip/wheels/4e/ca/bf/b020d2be95f7641801a6597a29c8f4f19e38f9c02a345bab9b
Successfully built fasttext


Основа для кода ниже взята из: https://towardsdatascience.com/fasttext-for-text-classification-a4b38cbff27c

In [180]:
# Saving the CSV file as a text file to train/test the classifier
train_sample = df.loc[:, [1, 2]]
train_sample.to_csv('train.txt', index = False, sep = ' ')


import fasttext
# Training the fastText classifier
model = fasttext.train_supervised('train.txt', wordNgrams = 2)

test_sample = df_test.loc[:, [1, 2]]
test_sample.to_csv('test1.txt', index = False, sep = ' ')
# Evaluating performance on the entire test file
n, pr, rec = model.test('test1.txt') 
print('Macro F1 score equals', 2 * (pr * rec) / (pr + rec))

Macro F1 score equals 0.1974713242961418


Таким образом, в качестве бенчмарка имеем F1 в размере 0.2

## 2. Самодельная нейросеть.

Нижеследующий код основан на https://coderzcolumn.com/tutorials/artificial-intelligence/pytorch-simple-guide-to-text-classification#2, https://towardsdatascience.com/lstm-text-classification-using-pytorch-2c6c657f8fc0

In [7]:
!pip install torchtext==0.10.1

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


Загрузим пакеты и предобработаем данные.

In [8]:
import torch
import torchtext
from torch.utils.data import DataLoader, Dataset

In [9]:
if torch.cuda.is_available():
    device = torch.device("cuda")
else:
    device = torch.device("cpu")

In [79]:
train_sample = df.loc[:, [1, 2]]
test_sample = df_test.loc[:, [1, 2]]

In [80]:
from sklearn.preprocessing import OrdinalEncoder
import numpy as np
enc = OrdinalEncoder()
enc.fit(np.array(train_sample.loc[:, 1]).reshape(-1, 1))
train_sample[1] = enc.transform(np.array(train_sample.loc[:, 1]).reshape(-1, 1))
test_sample[1] = enc.transform(np.array(test_sample.loc[:, 1]).reshape(-1, 1))

In [81]:
class MyDataset(Dataset):
 
  def __init__(self, train = True):
    if train:
      x=train_sample[2].values
      y=train_sample[1].values
    else:
      x=test_sample[2].values
      y=test_sample[1].values
 
    self.x_train=x
    self.y_train=torch.tensor(y).type(torch.LongTensor)
 
  def __len__(self):
    return len(self.y_train)
   
  def __getitem__(self,idx):
    return self.x_train[idx],self.y_train[idx]

In [82]:
train_dataset, test_dataset = MyDataset(), MyDataset(train=False)

Сделаем векторизацию батча при помощи Word2Vec эмбеддингов (каждый текст представим в виде среднего эмбеддингов входящих в него слов).

In [14]:
class Text2VecTransformer:
    def __init__(self, model_dir='./'):
        model = Word2Vec.load(model_dir + "hpac_w2v.model")
        self.model = model
        
    def transform(self, text: str):
        vecs = []
        for word in text.split():
            try:
                word_embedding = self.model.wv[word]
                vecs.append(word_embedding)
            except KeyError:  # если слова нет в словаре w2v, то пропускаем его
                continue
                
        vecs = np.array(vecs)
        return self.aggregation(vecs)
        
    def aggregation(self, vecs: np.array):
        return np.mean(vecs, axis=0)
#         return np.max(vecs, axis=0)  # max-pool ?
    
    def text_summary(self, text: str):
        """Для проверки на здравый смысл и поиска лучшей функции агрегации"""
        return model.wv.similar_by_vector(self.transform(text))

In [20]:
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from torchtext.data.functional import to_map_style_dataset

vectorizer = Text2VecTransformer()

def vectorize_batch(batch):
    X, Y = list(zip(*batch))
    X_vect = np.array([vectorizer.transform(x) for x in X])
    return torch.tensor(X_vect, dtype=torch.float32), torch.tensor(Y)

#train_dataset, test_dataset  = torchtext.datasets.AG_NEWS()
#train_dataset, test_dataset = to_map_style_dataset(train_dataset), to_map_style_dataset(test_dataset)

train_loader = DataLoader(train_dataset, batch_size=256, collate_fn=vectorize_batch)
test_loader  = DataLoader(test_dataset, batch_size=256, collate_fn=vectorize_batch)

Напишем классификатор, представляющий собой нейросеть, где сначала вход поступает в LSTM слой, за которым идут два линейных слоя. За основу взят код отсюда: https://towardsdatascience.com/lstm-text-classification-using-pytorch-2c6c657f8fc0

In [39]:
from torch import nn
from torch.nn import functional as F
from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence

class LSTM(nn.Module):

    def __init__(self, dimension=128):
        super(LSTM, self).__init__()

        #self.embedding = nn.Embedding(len(text_field.vocab), 300)
        self.dimension = dimension
        self.lstm = nn.LSTM(input_size=250,
                            hidden_size=dimension,
                            num_layers=1,
                            batch_first=True,
                            bidirectional=False).to(device)
        self.drop = nn.Dropout(p=0.05).to(device)

        self.fc1 = nn.Linear(128, 300).to(device)
        self.fc2 = nn.Linear(300, len(np.unique(train_sample[1]))).to(device)
        self.relu = nn.ReLU().to(device)

    def forward(self, X_batch):
        X_batch =  X_batch.reshape(-1,1, 250).to(device)
        #packed_input = pack_padded_sequence(fc1_out, text_len, batch_first=True, enforce_sorted=False)
        lstm_out, _ = self.lstm(X_batch)
        lstm_out = self.drop(lstm_out[:,-1])
        fc1_out = self.fc1(lstm_out)
        fc1_out = self.relu(fc1_out)
        fc2_out = self.fc2(fc1_out)
        return fc2_out

Попробуем классификатор CNN отсюда: https://towardsdatascience.com/text-classification-with-cnns-in-pytorch-1113df31e79f.

In [87]:
import math
from torch import nn
from torch.nn import functional as F
class CNN(nn.ModuleList):

    def __init__(self):
      super(CNN, self).__init__()

      # Parameters regarding text preprocessing
      self.seq_len = 250
      #self.num_words = params.num_words
      #self.embedding_size = params.embedding_size
      
      # Dropout definition
      self.dropout = nn.Dropout(0.25)
      
      # CNN parameters definition
      # Kernel sizes
      self.kernel_1 = 2
      self.kernel_2 = 3
      self.kernel_3 = 4
      self.kernel_4 = 5
      
      # Output size for each convolution
      self.out_size = 1
      # Number of strides for each convolution
      self.stride = 2
      
      # Embedding layer definition
      #self.embedding = nn.Embedding(self.num_words + 1, self.embedding_size, padding_idx=0)
      
      # Convolution layers definition
      self.conv_1 = nn.Conv1d(1, self.out_size, self.kernel_1, self.stride).to(device)
      self.conv_2 = nn.Conv1d(1, self.out_size, self.kernel_2, self.stride).to(device)
      self.conv_3 = nn.Conv1d(1, self.out_size, self.kernel_3, self.stride).to(device)
      self.conv_4 = nn.Conv1d(1, self.out_size, self.kernel_4, self.stride).to(device)
      
      # Max pooling layers definition
      self.pool_1 = nn.MaxPool1d(self.kernel_1, self.stride).to(device)
      self.pool_2 = nn.MaxPool1d(self.kernel_2, self.stride).to(device)
      self.pool_3 = nn.MaxPool1d(self.kernel_3, self.stride).to(device)
      self.pool_4 = nn.MaxPool1d(self.kernel_4, self.stride).to(device)
      
      # Fully connected layer definition
      self.fc = nn.Linear(244, len(np.unique(train_sample[1]))).to(device)

    def forward(self, X_batch):

      # Sequence of tokes is filterd through an embedding layer
      x = torch.reshape(X_batch, (-1, 1, self.seq_len)).to(device)
      
      # Convolution layer 1 is applied
      x1 = self.conv_1(x)
      x1 = torch.relu(x1)
      x1 = self.pool_1(x1)
      
      # Convolution layer 2 is applied
      x2 = self.conv_2(x)
      x2 = torch.relu((x2))
      x2 = self.pool_2(x2)
   
      # Convolution layer 3 is applied
      x3 = self.conv_3(x)
      x3 = torch.relu(x3)
      x3 = self.pool_3(x3)
      
      # Convolution layer 4 is applied
      x4 = self.conv_4(x)
      x4 = torch.relu(x4)
      x4 = self.pool_4(x4)
      
      # The output of each convolutional layer is concatenated into a unique vector
      union = torch.cat((x1, x2, x3, x4), 2)
      union = union.reshape(union.size(0), -1)

      # The "flattened" vector is passed through a fully connected layer
      out = self.fc(union)
      # Dropout is applied		
      #out = self.dropout(out)
      # Activation function is applied
      #out = torch.sigmoid(out)
      
      return out.squeeze()

Запрограммируем процесс обучения и тестирования.

In [35]:
from tqdm import tqdm
from sklearn.metrics import f1_score, accuracy_score
import gc

def CalcValLossAndAccuracy(model, loss_fn, val_loader):
    with torch.no_grad():
        Y_shuffled, Y_preds, losses = [],[],[]
        for X, Y in val_loader:
            preds = model(X)
            Y = Y.to(device)
            loss = loss_fn(preds, Y)
            losses.append(loss.item())

            Y_shuffled.append(Y)
            Y_preds.append(preds.argmax(dim=-1))

        Y_shuffled = torch.cat(Y_shuffled)
        Y_preds = torch.cat(Y_preds)

        print("Valid Loss : {:.3f}".format(torch.tensor(losses).mean()))
        print("Valid Macro F1  : {:.3f}".format(f1_score(Y_shuffled.detach().cpu(), Y_preds.detach().cpu(), average='macro')))


def TrainModel(model, loss_fn, optimizer, train_loader, val_loader, epochs=10):
    for i in range(1, epochs+1):
        losses = []
        for X, Y in tqdm(train_loader):
            Y_preds = model(X)
            Y = Y.to(device)

            loss = loss_fn(Y_preds, Y)
            losses.append(loss.item())

            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

        print("Train Loss : {:.3f}".format(torch.tensor(losses).mean()))
        CalcValLossAndAccuracy(model, loss_fn, val_loader)

Обучим LSTM.

In [40]:
from torch.optim import Adam

epochs = 10
learning_rate = 1e-3

loss_fn = nn.CrossEntropyLoss()
#text_classifier = MLP()
text_classifier = LSTM()
optimizer = Adam(text_classifier.parameters(), lr=learning_rate)

TrainModel(text_classifier, loss_fn, optimizer, train_loader, test_loader, epochs)

100%|██████████| 238/238 [00:19<00:00, 12.17it/s]


Train Loss : 3.200
Valid Loss : 2.954
Valid Macro F1  : 0.033


100%|██████████| 238/238 [00:19<00:00, 12.19it/s]


Train Loss : 2.879
Valid Loss : 2.836
Valid Macro F1  : 0.049


100%|██████████| 238/238 [00:22<00:00, 10.55it/s]


Train Loss : 2.789
Valid Loss : 2.781
Valid Macro F1  : 0.061


100%|██████████| 238/238 [00:19<00:00, 12.13it/s]


Train Loss : 2.736
Valid Loss : 2.751
Valid Macro F1  : 0.069


100%|██████████| 238/238 [00:19<00:00, 12.14it/s]


Train Loss : 2.696
Valid Loss : 2.737
Valid Macro F1  : 0.072


100%|██████████| 238/238 [00:21<00:00, 11.02it/s]


Train Loss : 2.663
Valid Loss : 2.728
Valid Macro F1  : 0.076


100%|██████████| 238/238 [00:19<00:00, 12.10it/s]


Train Loss : 2.634
Valid Loss : 2.716
Valid Macro F1  : 0.077


100%|██████████| 238/238 [00:19<00:00, 12.13it/s]


Train Loss : 2.608
Valid Loss : 2.715
Valid Macro F1  : 0.086


100%|██████████| 238/238 [00:23<00:00,  9.99it/s]


Train Loss : 2.582
Valid Loss : 2.711
Valid Macro F1  : 0.086


100%|██████████| 238/238 [00:19<00:00, 12.17it/s]


Train Loss : 2.558
Valid Loss : 2.708
Valid Macro F1  : 0.089


Обучим CNN

In [88]:
from torch.optim import Adam

epochs = 10
learning_rate = 1e-3

loss_fn = nn.CrossEntropyLoss()
text_classifier = CNN()
#text_classifier = LSTM()
optimizer = Adam(text_classifier.parameters(), lr=learning_rate)

TrainModel(text_classifier, loss_fn, optimizer, train_loader, test_loader, epochs)

100%|██████████| 277/277 [00:24<00:00, 11.49it/s]


Train Loss : 3.453
Valid Loss : 3.249
Valid Macro F1  : 0.014


100%|██████████| 277/277 [00:24<00:00, 11.47it/s]


Train Loss : 3.151
Valid Loss : 3.097
Valid Macro F1  : 0.027


100%|██████████| 277/277 [00:23<00:00, 11.57it/s]


Train Loss : 3.043
Valid Loss : 3.020
Valid Macro F1  : 0.035


100%|██████████| 277/277 [00:23<00:00, 11.58it/s]


Train Loss : 2.981
Valid Loss : 2.975
Valid Macro F1  : 0.040


100%|██████████| 277/277 [00:24<00:00, 11.53it/s]


Train Loss : 2.937
Valid Loss : 2.944
Valid Macro F1  : 0.045


100%|██████████| 277/277 [00:23<00:00, 11.57it/s]


Train Loss : 2.904
Valid Loss : 2.923
Valid Macro F1  : 0.048


100%|██████████| 277/277 [00:23<00:00, 11.71it/s]


Train Loss : 2.878
Valid Loss : 2.908
Valid Macro F1  : 0.051


100%|██████████| 277/277 [00:25<00:00, 10.71it/s]


Train Loss : 2.857
Valid Loss : 2.897
Valid Macro F1  : 0.057


100%|██████████| 277/277 [00:23<00:00, 11.72it/s]


Train Loss : 2.840
Valid Loss : 2.890
Valid Macro F1  : 0.058


100%|██████████| 277/277 [00:23<00:00, 11.59it/s]


Train Loss : 2.825
Valid Loss : 2.884
Valid Macro F1  : 0.063


Как видим, CNN оказалась хуже LSTM.

Мы также пробовали использовать one-hot эмбеддинги текстов для обучения (код для их получения приведен ниже), но качество  было лишь незначительно ниже.

In [ ]:
from torchtext.data import get_tokenizer
from torchtext.vocab import build_vocab_from_iterator

tokenizer = get_tokenizer("basic_english")

def build_vocab(datasets):
    for dataset in datasets:
        for text, label in dataset:
            yield tokenizer(text)

vocab = build_vocab_from_iterator(build_vocab([train_dataset, test_dataset]), specials=["<UNK>"])
vocab.set_default_index(vocab["<UNK>"])

from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from torchtext.data.functional import to_map_style_dataset

vectorizer = CountVectorizer(vocabulary=vocab.get_itos(), tokenizer=tokenizer)

def vectorize_batch(batch):
    X, Y = list(zip(*batch))
    X = vectorizer.transform(X).todense()
    return torch.tensor(X, dtype=torch.float32), torch.tensor(Y) #- 1 ## We have deducted 1 from target names to get them in range [0,1,2,3] from [1,2,3,4]

#train_dataset, test_dataset  = torchtext.datasets.AG_NEWS()
#train_dataset, test_dataset = to_map_style_dataset(train_dataset), to_map_style_dataset(test_dataset)

train_loader = DataLoader(train_dataset, batch_size=256, collate_fn=vectorize_batch)
test_loader  = DataLoader(test_dataset, batch_size=256, collate_fn=vectorize_batch)

## 3. Аугментации.

Дополним обучающее множество сэмплами, в каждом из которых 4 случайных слова заменены синонимами. Получим эти сэмплы, аугментировав первые 10000 наблюдений из обучающей выборки.

In [41]:
!pip install nlpaug

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 410 kB 5.0 MB/s 


In [66]:
train_sample = df.loc[:, [1, 2]]
test_sample = df_test.loc[:, [1, 2]]

In [67]:
from sklearn.preprocessing import OrdinalEncoder
import numpy as np
enc = OrdinalEncoder()
enc.fit(np.array(train_sample.loc[:, 1]).reshape(-1, 1))
train_sample[1] = enc.transform(np.array(train_sample.loc[:, 1]).reshape(-1, 1))
test_sample[1] = enc.transform(np.array(test_sample.loc[:, 1]).reshape(-1, 1))

In [68]:
#Augmentations 
import nltk
nltk.download('punkt')
import nlpaug.augmenter.sentence as nas
import nlpaug
import nlpaug.augmenter.word as naw

#Synonym Replacement
aug = naw.SynonymAug(aug_src='wordnet',aug_max=4)
aug_train = train_sample.loc[:10000, :]
aug_train.loc[:10000, 2] = train_sample.loc[:10000, 2].apply(lambda x: aug.augment(x)[0])

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


In [89]:
train_sample = pd.concat([train_sample, aug_train]).reset_index(drop=True)

In [71]:
class MyDataset(Dataset):
 
  def __init__(self, train = True):
    if train:
      x=train_sample[2].values
      y=train_sample[1].values
    else:
      x=test_sample[2].values
      y=test_sample[1].values
 
    self.x_train=x
    self.y_train=torch.tensor(y).type(torch.LongTensor)
 
  def __len__(self):
    return len(self.y_train)
   
  def __getitem__(self,idx):
    return self.x_train[idx],self.y_train[idx]

In [72]:
train_dataset, test_dataset = MyDataset(), MyDataset(train=False)

In [73]:
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
#from torchtext.data.functional import to_map_style_dataset

vectorizer = Text2VecTransformer()

def vectorize_batch(batch):
    X, Y = list(zip(*batch))
    X_vect = np.array([vectorizer.transform(x) for x in X])
    return torch.tensor(X_vect, dtype=torch.float32), torch.tensor(Y)

#train_dataset, test_dataset  = torchtext.datasets.AG_NEWS()
#train_dataset, test_dataset = to_map_style_dataset(train_dataset), to_map_style_dataset(test_dataset)

train_loader = DataLoader(train_dataset, batch_size=256, collate_fn=vectorize_batch)
test_loader  = DataLoader(test_dataset, batch_size=256, collate_fn=vectorize_batch)

Протестируем на лучшей нейросети LSTM.

In [74]:
from torch.optim import Adam

epochs = 10
learning_rate = 1e-3

loss_fn = nn.CrossEntropyLoss()
#text_classifier = MLP()
text_classifier = LSTM()
optimizer = Adam(text_classifier.parameters(), lr=learning_rate)

TrainModel(text_classifier, loss_fn, optimizer, train_loader, test_loader, epochs)

100%|██████████| 277/277 [00:23<00:00, 11.58it/s]


Train Loss : 3.154
Valid Loss : 2.914
Valid Macro F1  : 0.035


100%|██████████| 277/277 [00:24<00:00, 11.46it/s]


Train Loss : 2.851
Valid Loss : 2.805
Valid Macro F1  : 0.056


100%|██████████| 277/277 [00:23<00:00, 11.98it/s]


Train Loss : 2.764
Valid Loss : 2.763
Valid Macro F1  : 0.065


100%|██████████| 277/277 [00:22<00:00, 12.11it/s]


Train Loss : 2.712
Valid Loss : 2.741
Valid Macro F1  : 0.071


100%|██████████| 277/277 [00:22<00:00, 12.14it/s]


Train Loss : 2.671
Valid Loss : 2.729
Valid Macro F1  : 0.075


100%|██████████| 277/277 [00:22<00:00, 12.17it/s]


Train Loss : 2.635
Valid Loss : 2.724
Valid Macro F1  : 0.076


100%|██████████| 277/277 [00:23<00:00, 12.04it/s]


Train Loss : 2.604
Valid Loss : 2.719
Valid Macro F1  : 0.080


100%|██████████| 277/277 [00:23<00:00, 12.00it/s]


Train Loss : 2.572
Valid Loss : 2.716
Valid Macro F1  : 0.085


100%|██████████| 277/277 [00:22<00:00, 12.13it/s]


Train Loss : 2.542
Valid Loss : 2.723
Valid Macro F1  : 0.085


100%|██████████| 277/277 [00:22<00:00, 12.07it/s]


Train Loss : 2.512
Valid Loss : 2.726
Valid Macro F1  : 0.089


Простестируем CNN.

In [90]:
from torch.optim import Adam

epochs = 10
learning_rate = 1e-3

loss_fn = nn.CrossEntropyLoss()
text_classifier = CNN()
#text_classifier = LSTM()
optimizer = Adam(text_classifier.parameters(), lr=learning_rate)

TrainModel(text_classifier, loss_fn, optimizer, train_loader, test_loader, epochs)

100%|██████████| 277/277 [00:26<00:00, 10.56it/s]


Train Loss : 3.408
Valid Loss : 3.192
Valid Macro F1  : 0.019


100%|██████████| 277/277 [00:23<00:00, 11.80it/s]


Train Loss : 3.105
Valid Loss : 3.073
Valid Macro F1  : 0.027


100%|██████████| 277/277 [00:25<00:00, 10.77it/s]


Train Loss : 3.022
Valid Loss : 3.020
Valid Macro F1  : 0.035


100%|██████████| 277/277 [00:23<00:00, 11.67it/s]


Train Loss : 2.974
Valid Loss : 2.987
Valid Macro F1  : 0.041


100%|██████████| 277/277 [00:26<00:00, 10.63it/s]


Train Loss : 2.938
Valid Loss : 2.962
Valid Macro F1  : 0.046


100%|██████████| 277/277 [00:23<00:00, 11.69it/s]


Train Loss : 2.909
Valid Loss : 2.943
Valid Macro F1  : 0.049


100%|██████████| 277/277 [00:25<00:00, 10.71it/s]


Train Loss : 2.884
Valid Loss : 2.928
Valid Macro F1  : 0.055


100%|██████████| 277/277 [00:23<00:00, 11.74it/s]


Train Loss : 2.862
Valid Loss : 2.916
Valid Macro F1  : 0.056


100%|██████████| 277/277 [00:23<00:00, 11.67it/s]


Train Loss : 2.844
Valid Loss : 2.907
Valid Macro F1  : 0.058


100%|██████████| 277/277 [00:25<00:00, 10.77it/s]


Train Loss : 2.828
Valid Loss : 2.899
Valid Macro F1  : 0.060


# Часть 4. Итоги
Подведем итоги:
* Пункт 1. Данный пункт представлял собой проведение эксплоративного анализа, в частности, здесь нами был произведен поиск наиболее часто употребляемых слов (п. 1.1.), а также имен, пар имя + фамилия, пар вида "профессор" + имя / фамилия (п 1.2.)
  * Пункт 1.1. Касательно результатов этого пункта можно сказать, что в целом картина по топ-словам является более-менее ожидаемой (с учетом тематики текстов, что отражается в основном в присутствии в топе имен главных персонажей). Немного неожиданным оказалось лишь то, что отрицательная частица "not" в сокращенной форме ("nt") является настолько часто употребляемой.
  * Пункт 1.2. Для поиска имен был использован внешний источник с данными о подавляющем большинстве персонажей вселенной Гарри Поттера. Процесс извлечения имен может быть распараллелен, так как чтение и проверка вхождения не зависят в текстах друг от друга. Для разделения на имя и фамилию отдельную сложность представляли ФИО с длиной более двух, например Thomas Marvolo Riddle или даже более длинные. Кроме того, отдельной обработки требовали такие имена, как Grey Lady, которые состоят двух слов, но при простом сплите по пробелу разделились бы на имя и фамилию неверно и могли значительно исказить итоговую статистику по частотности. Для решения упомянутых сложностей предобработка списка имен производилась в ручном режиме. В конечном итоге, по всем трем подпунктам получились ожидаемые результаты - это центральные персонажи.
* Пункт 2. В данном пункте на предоставленном корпусе нами была обучена модель word2vec. Проанализировав результаты поиска синонимов, ассоциаций и лишних слов с использованием обученной модели, а также визуализацию на плоскости t-SNE представлений слов из п. 1.1, мы заключили, что полученная модель является адекватной.  
Небольшая техническая ремарка про трейд-офф между оперативной памятью и скоростью работы. Использование генераторов и чтение одного документа за раз позволяет снизить нагрузку на оперативную память, но при этом скорость обучения word2vec низкая, так как нет параллельности. Для использования многопоточности необходима предобработка данных, заключающаяся в записи единого обучающего корпуса, где каждый документ располагается на новой строке. Такой подход позволяет использовать многопоточность в процессе обучения, но тогда при формирования корпуса мы работаем в один поток. 
* Пункт 3.
  * Пункт 3.1. В качестве бенчмарк классификатора был использова Fasttext. Качество классификации (F1 score на тесте) составило около 0.2.
  * Пункт 3.2. Нами была обучены CNN и LSTM сети для классификации текстов. LSTM показала себя лучше CNN. Мы обучали сети на полученных в предыдущих пунктах word2vec эмбеддингах. F1 score на тесте для LSTM после 10 эпох составило около 0.09, что, хотя и ниже нашего бейзлайна из пункта 3.1, вполне соответсвует результатам полученным для подобной модели в базовой статье (см. табл. 3): https://arxiv.org/pdf/1905.11037.pdf
  * Пункт 3.3. Добавление аугментированных сэмплов не улучшило качества модели. Возможно, стоит попробовать другие типа аугментации (напр., удаление случайного слова) или увеличить количество аугментированных сэмплов.